src/popover-helpers.ts

```diff
@@ -354,10 +454,62 @@ function closeAllOpenPopovers(
   focusPreviousElement = false,
   fireEvents = false,
 ) {
-  let popover = topMostAutoPopover(document);
+  let popover = topMostAutoOrHintPopover(document);
   while (popover) {
     hidePopover(popover, focusPreviousElement, fireEvents);
-    popover = topMostAutoPopover(document);
+    popover = topMostAutoOrHintPopover(document);
+  }
+}
+
+function closeAllOpenPopoversInList(
+  list: Set<HTMLElement>,
+  focusPreviousElement = false,
+  fireEvents = false,
+) {
+  let popover = topMostPopoverInList(list);
+  while (popover) {
+    hidePopover(popover, focusPreviousElement, fireEvents);
+    popover = topMostPopoverInList(list);
+  }
+}
+
+// https://html.spec.whatwg.org/#hide-popover-stack-until
+function hidePopoverStackUntil(
+  endpoint: Element | Document,
+  set: Set<HTMLElement>,
+  focusPreviousElement: boolean,
+  fireEvents: boolean,
+) {
+  let repeatingHide = false;
+  let hasRunOnce = false;
+  // TODO: This doesn't match the spec changes in
+  // https://github.com/whatwg/html/pull/9778/files#diff-41cf6794ba4200b839c53531555f0f3998df4cbb01a4d5cb0b94e3ca5e23947dR85830
+  while (repeatingHide || !hasRunOnce) {
+    hasRunOnce = true;
+    let lastToHide = null;
+    let foundEndpoint = false;
+    for (const popover of set) {
+      if (popover === endpoint) {
+        foundEndpoint = true;
+      } else if (foundEndpoint) {
+        lastToHide = popover;
+        break;
+      }
+    }
```

Comment:
I went down a rabbit hole, and this really has me stumped. The spec logic was updated in such a way that AFAICT sets `lastToHide` as the element _before_ the endpoint, whereas the prior logic set `lastToHide` as the element `after` the endpoint. The way I understand the spec, this section should be:

```suggestion
    let lastToHide = null;
    for (const popover of set) {
      if (popover === endpoint) {
        break;
      }
      lastToHide = popover;
    }
```

The problem with that is it seems to cause an infinite recursion error, because when we call `hidePopover` on the last element in the set, we end up finding the exact same `lastToHide` and calling `hidePopover` again, etc.

Is this a bug in the spec, or are we misreading it, or...? 😕

Reply:
It looks like it is this spec bug- https://github.com/whatwg/html/issues/11007. @jgerigmeyer does that match what you were seeing?

Reply:
Yes, exactly. We are passing the set by reference so that isn't an issue for our implementation, but the loop-reverse bug is what we were hitting. Based on that issue, it seems like our current implementation is fine — at least until they change the spec. I mean we could use the spec logic and also reverse the loop, but I don’t see how that really changes anything.